jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/FilesTaskV2.java

```diff
@@ -0,0 +1,193 @@
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskAction;
+
+/**
+ * Print out changing source dependencies on a project.
+ *
+ * <p>Expected use: {@code ./gradlew _jibSkaffoldFiles -q} or {@code ./gradlew
+ * :<subproject>:_jibSkaffoldFiles -q}
+ */
+public class FilesTaskV2 extends DefaultTask {
+
+  private final SkaffoldFilesOutput skaffoldFilesOutput = new SkaffoldFilesOutput();
+
+  @Nullable private JibExtension jibExtension;
+
+  public FilesTaskV2 setJibExtension(JibExtension jibExtension) {
+    this.jibExtension = jibExtension;
+    return this;
+  }
+
+  @TaskAction
+  public void listFiles() throws IOException {
+    Preconditions.checkNotNull(jibExtension);
+    Project project = getProject();
+
+    // If this is not the root project, add the root project's build.gradle and settings.gradle
+    if (project != project.getRootProject()) {
+      addGradleFiles(project.getRootProject());
+    }
+
+    addProjectFiles(project);
+
+    // Add extra layer
+    if (Files.exists(jibExtension.getExtraDirectory().getPath())) {
+      skaffoldFilesOutput.addInput(jibExtension.getExtraDirectory().getPath());
+    }
+
+    // Find project dependencies
+    Set<ProjectDependency> projectDependencies = findProjectDependencies(project);
+
+    Set<File> projectDependencyJars = new HashSet<>();
+    for (ProjectDependency projectDependency : projectDependencies) {
+      addProjectFiles(projectDependency.getDependencyProject());
+
+      // Keep track of project dependency jars for filtering out later
+      String configurationName = projectDependency.getTargetConfiguration();
+      if (configurationName == null) {
+        configurationName = "default";
+      }
+      Project dependencyProject = projectDependency.getDependencyProject();
+      for (Configuration targetConfiguration :
+          dependencyProject.getConfigurations().getByName(configurationName).getHierarchy()) {
+        for (PublishArtifact artifact : targetConfiguration.getArtifacts()) {
+          projectDependencyJars.add(artifact.getFile());
+        }
+      }
+    }
+
+    // Add SNAPSHOT, non-project dependency jars
+    for (File file : project.getConfigurations().getByName("runtime")) {
+      if (!projectDependencyJars.contains(file) && file.toString().contains("SNAPSHOT")) {
+        skaffoldFilesOutput.addInput(file.toPath());
+        projectDependencyJars.add(file); // Add to set to avoid printing the same files twice
+      }
+    }
+
+    // Print files
+    System.out.println(skaffoldFilesOutput.getJsonString());
+  }
+
+  /**
+   * Prints the locations of a project's build.gradle, settings.gradle, and gradle.properties.
+   *
+   * @param project the project
+   */
+  private void addGradleFiles(Project project) {
```

Comment:
we may also need to deal with script plugins here (https://docs.gradle.org/current/userguide/plugins.html#sec:script_plugins)

I think there's a way to query it, but I'm not sure.

Reply:
maybe a TODO, presumably this isn't available in v1 either

Reply:
filed: https://discuss.gradle.org/t/programmatically-access-script-plugin-references/30668 for now, lets see what they say.

Reply:
Hmm, yeah I'll open an issue for this.

Reply:
https://github.com/GoogleContainerTools/jib/issues/1494